File: examples/lightning_operations.py
#!/usr/bin/env python3
"""
Complete gBTC Lightning Operations Examples

This example demonstrates:
1. gBTC Lift (On-Ramp): Lightning → VTXO conversion
2. gBTC Land (Off-Ramp): VTXO → Lightning conversion
3. Complete Nostr event flows for Lightning operations
4. Error handling and status monitoring

Usage:
    python lightning_operations.py --gateway http://localhost:8000 lift --amount 100000
    python lightning_operations.py --gateway http://localhost:8000 land --amount 50000 --invoice lnbc...
    python lightning_operations.py --gateway http://localhost:8000 monitor --session-id sess_123
"""

from __future__ import annotations

import argparse
import json
import os
import sys
import time
from pathlib import Path
from typing import Any, Dict, Optional

# Ensure we can import the SDK from the repo root
REPO_ROOT = Path(__file__).resolve().parents[2]
if str(REPO_ROOT) not in sys.path:
    sys.path.insert(0, str(REPO_ROOT))

from sdk.gateway_client import GatewayClient, GatewayClientError  # noqa: E402


class LightningOperationsDemo:
    """Complete Lightning operations demo with Nostr event flows"""

    def __init__(self, gateway_url: str):
        self.client = GatewayClient(gateway_url)
        self.gateway_npub = "npub1gateway..."  # Replace with actual gateway npub

    def create_lift_intent(self, amount_sats: int, asset_id: str = "gBTC") -> Dict[str, Any]:
        """Create a 31510 intent for gBTC lift operation"""
        return {
            "action_id": f"lift_{int(time.time())}_{os.urandom(4).hex()}",
            "type": "lightning:lift",
            "params": {
                "asset_id": asset_id,
                "amount": amount_sats,
                "fee_asset_id": "gBTC",
                "fee_amount": 10  # 10 sats in gBTC
            },
            "expires_at": int(time.time()) + 15 * 60  # 15 minutes
        }

    def create_land_intent(self, amount_sats: int, lightning_invoice: str, asset_id: str = "gBTC") -> Dict[str, Any]:
        """Create a 31510 intent for gBTC land operation"""
        return {
            "action_id": f"land_{int(time.time())}_{os.urandom(4).hex()}",
            "type": "lightning:land",
            "params": {
                "asset_id": asset_id,
                "amount": amount_sats,
                "lightning_invoice": lightning_invoice,
                "fee_asset_id": "gBTC",
                "fee_amount": int(amount_sats * 0.001)  # 0.1% fee
            },
            "expires_at": int(time.time()) + 15 * 60
        }

    def simulate_31500_service_request(self, action: str, asset_id: str, amount: int) -> Dict[str, Any]:
        """Simulate 31500 Service Request for Lightning operations"""
        request = {
            "action": action,
            "asset_id": asset_id,
            "amount": amount,
            "timestamp": int(time.time())
        }

        print(f"\n📤 31500 Service Request:")
        print(json.dumps(request, indent=2))
        return request

    def simulate_31501_service_response(self, request: Dict[str, Any], invoice: Optional[str] = None) -> Dict[str, Any]:
        """Simulate 31501 Service Response from gateway"""
        response = {
            "status": "pending",
            "action": request["action"],
            "asset_id": request["asset_id"],
            "amount": request["amount"],
            "timestamp": int(time.time())
        }

        if request["action"] == "lift_lightning" and invoice:
            response["invoice"] = invoice
            response["expires_at"] = int(time.time()) + 3600  # 1 hour

        print(f"\n📥 31501 Service Response:")
        print(json.dumps(response, indent=2))
        return response

    def execute_lift_flow(self, amount_sats: int, asset_id: str = "gBTC") -> Dict[str, Any]:
        """Complete gBTC lift flow: Lightning → VTXO"""
        print(f"\n🚀 Starting gBTC Lift Flow: {amount_sats} sats")

        # Step 1: 31500 Service Request
        service_request = self.simulate_31500_service_request("lift_lightning", asset_id, amount_sats)

        # Step 2: Get Lightning invoice from gateway (simulated)
        mock_invoice = f"lnbc{amount_sats}n1p3k8..."  # Mock invoice
        service_response = self.simulate_31501_service_response(service_request, mock_invoice)

        # Step 3: Create 31510 intent (after user pays Lightning invoice)
        intent = self.create_lift_intent(amount_sats, asset_id)
        print(f"\n📋 31510 Intent (after Lightning payment):")
        print(json.dumps(intent, indent=2))

        # Step 4: Create session
        session_resp = self.client.create_session(
            user_pubkey="npub1user...",  # Replace with actual user npub
            session_type="lightning_lift",
            intent_data=intent
        )
        session_id = session_resp.get("session_id")
        print(f"\n✅ Session created: {session_id}")

        # Step 5: Create challenge
        challenge_data = {
            "payload_to_sign": f"0x{hashlib.sha256(json.dumps(intent, sort_keys=True).encode()).hexdigest()}",
            "type": "sign_payload",
            "payload_ref": f"sha256:{hashlib.sha256(json.dumps(intent['params'], sort_keys=True).encode()).hexdigest()}"
        }
        context = {
            "human": f"Authorize gBTC lift of {amount_sats} sats via Lightning payment",
            "step_index": 1,
            "step_total": 1
        }

        challenge = self.client.create_challenge(
            session_id=session_id,
            challenge_data=challenge_data,
            context=context
        )
        print(f"\n🔐 Challenge created: {challenge.get('challenge_id')}")

        # Step 6: Start ceremony (simulating user approval)
        ceremony_resp = self.client.start_ceremony(session_id=session_id)
        print(f"\n🎭 Ceremony started: {ceremony_resp.get('status')}")

        return {
            "session_id": session_id,
            "intent": intent,
            "service_request": service_request,
            "service_response": service_response,
            "status": "lift_initiated"
        }

    def execute_land_flow(self, amount_sats: int, lightning_invoice: str, asset_id: str = "gBTC") -> Dict[str, Any]:
        """Complete gBTC land flow: VTXO → Lightning"""
        print(f"\n🛬 Starting gBTC Land Flow: {amount_sats} sats")

        # Step 1: Create 31510 intent
        intent = self.create_land_intent(amount_sats, lightning_invoice, asset_id)
        print(f"\n📋 31510 Intent:")
        print(json.dumps(intent, indent=2))

        # Step 2: Create session
        session_resp = self.client.create_session(
            user_pubkey="npub1user...",  # Replace with actual user npub
            session_type="lightning_land",
            intent_data=intent
        )
        session_id = session_resp.get("session_id")
        print(f"\n✅ Session created: {session_id}")

        # Step 3: Create challenge for VTXO spending
        challenge_data = {
            "payload_to_sign": f"0x{hashlib.sha256(json.dumps(intent, sort_keys=True).encode()).hexdigest()}",
            "type": "sign_tx",
            "payload_ref": f"sha256:{hashlib.sha256(json.dumps(intent['params'], sort_keys=True).encode()).hexdigest()}"
        }
        context = {
            "human": f"Authorize gBTC land of {amount_sats} sats to Lightning invoice",
            "step_index": 1,
            "step_total": 1
        }

        challenge = self.client.create_challenge(
            session_id=session_id,
            challenge_data=challenge_data,
            context=context
        )
        print(f"\n🔐 Challenge created: {challenge.get('challenge_id')}")

        # Step 4: Start ceremony
        ceremony_resp = self.client.start_ceremony(session_id=session_id)
        print(f"\n🎭 Ceremony started: {ceremony_resp.get('status')}")

        return {
            "session_id": session_id,
            "intent": intent,
            "status": "land_initiated"
        }

    def monitor_session(self, session_id: str, timeout: int = 120) -> Dict[str, Any]:
        """Monitor session status and print 31340/31341 events"""
        print(f"\n📊 Monitoring session: {session_id}")

        start_time = time.time()
        while time.time() - start_time < timeout:
            try:
                status = self.client.get_ceremony_status(session_id=session_id)
                print(f"📈 Status: {status.get('status')} | Step: {status.get('current_step', 'N/A')}")

                if status.get('status') in ['completed', 'failed']:
                    print(f"\n🎉 Session {status.get('status')}!")

                    # Simulate 31340/31341 event
                    if status.get('status') == 'completed':
                        event_31340 = {
                            "status": "success",
                            "ref_action_id": status.get('intent_data', {}).get('action_id'),
                            "results": {
                                "txid": "mock_txid_abc123...",
                                "amount_processed": status.get('intent_data', {}).get('params', {}).get('amount'),
                                "fee_paid": status.get('intent_data', {}).get('params', {}).get('fee_amount')
                            },
                            "timestamp": int(time.time())
                        }
                        print(f"\n✅ 31340 Success Event:")
                        print(json.dumps(event_31340, indent=2))
                    else:
                        event_31341 = {
                            "status": "failure",
                            "code": status.get('error_code', 1000),
                            "message": status.get('error_message', 'Unknown error'),
                            "ref_action_id": status.get('intent_data', {}).get('action_id'),
                            "timestamp": int(time.time())
                        }
                        print(f"\n❌ 31341 Failure Event:")
                        print(json.dumps(event_31341, indent=2))

                    return status

                time.sleep(5)

            except GatewayClientError as e:
                print(f"❌ Error monitoring session: {e}")
                break

        print(f"\n⏰ Timeout waiting for session completion")
        return {"status": "timeout"}


def main():
    parser = argparse.ArgumentParser(description="gBTC Lightning Operations Demo")
    parser.add_argument("--gateway", default=os.getenv("GATEWAY_BASE_URL", "http://localhost:8000"),
                       help="Gateway base URL")
    parser.add_argument("--user-npub", required=True, help="User's npub key")

    subparsers = parser.add_subparsers(dest="command", help="Operation to perform")

    # Lift command
    lift_parser = subparsers.add_parser("lift", help="gBTC lift operation")
    lift_parser.add_argument("--amount", type=int, required=True, help="Amount in satoshis")
    lift_parser.add_argument("--asset", default="gBTC", help="Asset ID (default: gBTC)")

    # Land command
    land_parser = subparsers.add_parser("land", help="gBTC land operation")
    land_parser.add_argument("--amount", type=int, required=True, help="Amount in satoshis")
    land_parser.add_argument("--invoice", required=True, help="Lightning invoice")
    land_parser.add_argument("--asset", default="gBTC", help="Asset ID (default: gBTC)")

    # Monitor command
    monitor_parser = subparsers.add_parser("monitor", help="Monitor session status")
    monitor_parser.add_argument("--session-id", required=True, help="Session ID to monitor")
    monitor_parser.add_argument("--timeout", type=int, default=120, help="Timeout in seconds")

    args = parser.parse_args()

    if not args.command:
        parser.print_help()
        return

    demo = LightningOperationsDemo(args.gateway)

    try:
        if args.command == "lift":
            result = demo.execute_lift_flow(args.amount, args.asset)
            print(f"\n🎯 Lift flow initiated. Session ID: {result['session_id']}")

        elif args.command == "land":
            result = demo.execute_land_flow(args.amount, args.invoice, args.asset)
            print(f"\n🎯 Land flow initiated. Session ID: {result['session_id']}")

        elif args.command == "monitor":
            result = demo.monitor_session(args.session_id, args.timeout)
            print(f"\n📊 Final status: {result.get('status')}")

    except GatewayClientError as e:
        print(f"❌ Gateway Error: {e}")
    except Exception as e:
        print(f"❌ Error: {e}")


if __name__ == "__main__":
    import hashlib  # For demo purposes
    main()

File: sdk-ts/examples/lightning-operations.ts
/**
 * Lightning Operations Examples for ArkRelay TypeScript SDK
 *
 * Demonstrates complete gBTC lift/land operations with Nostr event flows
 *
 * Usage:
 * import { LightningOperations } from './lightning-operations';
 * const lightning = new LightningOperations('http://localhost:8000');
 *
 * // Lift flow
 * const liftResult = await lightning.executeLiftFlow(100000, 'gBTC');
 *
 * // Land flow
 * const landResult = await lightning.executeLandFlow(50000, 'lnbc...');
 */

import { GatewayClient, GatewayClientError } from '../src';
import { createHash, randomBytes } from 'crypto';

export interface LightningIntent {
  actionId: string;
  type: string;
  params: {
    assetId: string;
    amount: number;
    feeAssetId: string;
    feeAmount: number;
  };
  expiresAt: number;
}

export interface SessionResult {
  sessionId: string;
  intent: LightningIntent;
  status: string;
}

export interface ServiceRequest {
  action: string;
  assetId: string;
  amount: number;
  timestamp: number;
}

export interface ServiceResponse {
  status: string;
  action: string;
  assetId: string;
  amount: number;
  timestamp: number;
  invoice?: string;
  expiresAt?: number;
}

export class LightningOperations {
  private client: GatewayClient;
  private gatewayNpub: string;

  constructor(gatewayUrl: string) {
    this.client = new GatewayClient(gatewayUrl);
    this.gatewayNpub = 'npub1gateway...'; // Replace with actual gateway npub
  }

  /**
   * Create a 31510 intent for gBTC lift operation
   */
  createLiftIntent(amountSats: number, assetId: string = 'gBTC'): LightningIntent {
    return {
      actionId: `lift_${Date.now()}_${randomBytes(4).toString('hex')}`,
      type: 'lightning:lift',
      params: {
        assetId,
        amount: amountSats,
        feeAssetId: 'gBTC',
        feeAmount: 10 // 10 sats in gBTC
      },
      expiresAt: Math.floor(Date.now() / 1000) + 15 * 60 // 15 minutes
    };
  }

  /**
   * Create a 31510 intent for gBTC land operation
   */
  createLandIntent(
    amountSats: number,
    lightningInvoice: string,
    assetId: string = 'gBTC'
  ): LightningIntent {
    return {
      actionId: `land_${Date.now()}_${randomBytes(4).toString('hex')}`,
      type: 'lightning:land',
      params: {
        assetId,
        amount: amountSats,
        lightningInvoice,
        feeAssetId: 'gBTC',
        feeAmount: Math.floor(amountSats * 0.001) // 0.1% fee
      },
      expiresAt: Math.floor(Date.now() / 1000) + 15 * 60
    };
  }

  /**
   * Simulate 31500 Service Request for Lightning operations
   */
  async simulate31500ServiceRequest(
    action: string,
    assetId: string,
    amount: number
  ): Promise<ServiceRequest> {
    const request: ServiceRequest = {
      action,
      assetId,
      amount,
      timestamp: Math.floor(Date.now() / 1000)
    };

    console.log('\n📤 31500 Service Request:');
    console.log(JSON.stringify(request, null, 2));
    return request;
  }

  /**
   * Simulate 31501 Service Response from gateway
   */
  async simulate31501ServiceResponse(
    request: ServiceRequest,
    invoice?: string
  ): Promise<ServiceResponse> {
    const response: ServiceResponse = {
      status: 'pending',
      action: request.action,
      assetId: request.assetId,
      amount: request.amount,
      timestamp: Math.floor(Date.now() / 1000)
    };

    if (request.action === 'lift_lightning' && invoice) {
      response.invoice = invoice;
      response.expiresAt = Math.floor(Date.now() / 1000) + 3600; // 1 hour
    }

    console.log('\n📥 31501 Service Response:');
    console.log(JSON.stringify(response, null, 2));
    return response;
  }

  /**
   * Complete gBTC lift flow: Lightning → VTXO
   */
  async executeLiftFlow(
    amountSats: number,
    assetId: string = 'gBTC'
  ): Promise<SessionResult> {
    console.log(`\n🚀 Starting gBTC Lift Flow: ${amountSats} sats`);

    try {
      // Step 1: 31500 Service Request
      const serviceRequest = await this.simulate31500ServiceRequest(
        'lift_lightning',
        assetId,
        amountSats
      );

      // Step 2: Get Lightning invoice from gateway (simulated)
      const mockInvoice = `lnbc${amountSats}n1p3k8...`; // Mock invoice
      const serviceResponse = await this.simulate31501ServiceResponse(
        serviceRequest,
        mockInvoice
      );

      // Step 3: Create 31510 intent (after user pays Lightning invoice)
      const intent = this.createLiftIntent(amountSats, assetId);
      console.log('\n📋 31510 Intent (after Lightning payment):');
      console.log(JSON.stringify(intent, null, 2));

      // Step 4: Create session
      const sessionResp = await this.client.createSession({
        userPubkey: 'npub1user...', // Replace with actual user npub
        sessionType: 'lightning_lift',
        intentData: intent
      });
      const sessionId = sessionResp.sessionId;
      console.log(`\n✅ Session created: ${sessionId}`);

      // Step 5: Create challenge
      const challengeData = {
        payloadToSign: `0x${createHash('sha256')
          .update(JSON.stringify(intent, Object.keys(intent).sort()))
          .digest('hex')}`,
        type: 'sign_payload',
        payloadRef: `sha256:${createHash('sha256')
          .update(JSON.stringify(intent.params, Object.keys(intent.params).sort()))
          .digest('hex')}`
      };

      const context = {
        human: `Authorize gBTC lift of ${amountSats} sats via Lightning payment`,
        stepIndex: 1,
        stepTotal: 1
      };

      const challenge = await this.client.createChallenge(sessionId, {
        challengeData,
        context
      });
      console.log(`\n🔐 Challenge created: ${challenge.challengeId}`);

      // Step 6: Start ceremony (simulating user approval)
      const ceremonyResp = await this.client.startCeremony(sessionId);
      console.log(`\n🎭 Ceremony started: ${ceremonyResp.status}`);

      return {
        sessionId,
        intent,
        status: 'lift_initiated'
      };

    } catch (error) {
      if (error instanceof GatewayClientError) {
        throw new Error(`Gateway Error: ${error.message}`);
      }
      throw error;
    }
  }

  /**
   * Complete gBTC land flow: VTXO → Lightning
   */
  async executeLandFlow(
    amountSats: number,
    lightningInvoice: string,
    assetId: string = 'gBTC'
  ): Promise<SessionResult> {
    console.log(`\n🛬 Starting gBTC Land Flow: ${amountSats} sats`);

    try {
      // Step 1: Create 31510 intent
      const intent = this.createLandIntent(amountSats, lightningInvoice, assetId);
      console.log('\n📋 31510 Intent:');
      console.log(JSON.stringify(intent, null, 2));

      // Step 2: Create session
      const sessionResp = await this.client.createSession({
        userPubkey: 'npub1user...', // Replace with actual user npub
        sessionType: 'lightning_land',
        intentData: intent
      });
      const sessionId = sessionResp.sessionId;
      console.log(`\n✅ Session created: ${sessionId}`);

      // Step 3: Create challenge for VTXO spending
      const challengeData = {
        payloadToSign: `0x${createHash('sha256')
          .update(JSON.stringify(intent, Object.keys(intent).sort()))
          .digest('hex')}`,
        type: 'sign_tx',
        payloadRef: `sha256:${createHash('sha256')
          .update(JSON.stringify(intent.params, Object.keys(intent.params).sort()))
          .digest('hex')}`
      };

      const context = {
        human: `Authorize gBTC land of ${amountSats} sats to Lightning invoice`,
        stepIndex: 1,
        stepTotal: 1
      };

      const challenge = await this.client.createChallenge(sessionId, {
        challengeData,
        context
      });
      console.log(`\n🔐 Challenge created: ${challenge.challengeId}`);

      // Step 4: Start ceremony
      const ceremonyResp = await this.client.startCeremony(sessionId);
      console.log(`\n🎭 Ceremony started: ${ceremonyResp.status}`);

      return {
        sessionId,
        intent,
        status: 'land_initiated'
      };

    } catch (error) {
      if (error instanceof GatewayClientError) {
        throw new Error(`Gateway Error: ${error.message}`);
      }
      throw error;
    }
  }

  /**
   * Monitor session status and print 31340/31341 events
   */
  async monitorSession(
    sessionId: string,
    timeout: number = 120000 // 2 minutes
  ): Promise<any> {
    console.log(`\n📊 Monitoring session: ${sessionId}`);

    const startTime = Date.now();

    return new Promise((resolve, reject) => {
      const interval = setInterval(async () => {
        try {
          const status = await this.client.getCeremonyStatus(sessionId);
          console.log(`📈 Status: ${status.status} | Step: ${status.currentStep || 'N/A'}`);

          if (status.status === 'completed' || status.status === 'failed') {
            clearInterval(interval);

            console.log(`\n🎉 Session ${status.status}!`);

            // Simulate 31340/31341 event
            if (status.status === 'completed') {
              const event31340 = {
                status: 'success',
                refActionId: status.intentData?.actionId,
                results: {
                  txid: 'mock_txid_abc123...',
                  amountProcessed: status.intentData?.params?.amount,
                  feePaid: status.intentData?.params?.feeAmount
                },
                timestamp: Math.floor(Date.now() / 1000)
              };
              console.log('\n✅ 31340 Success Event:');
              console.log(JSON.stringify(event31340, null, 2));
            } else {
              const event31341 = {
                status: 'failure',
                code: status.errorCode || 1000,
                message: status.errorMessage || 'Unknown error',
                refActionId: status.intentData?.actionId,
                timestamp: Math.floor(Date.now() / 1000)
              };
              console.log('\n❌ 31341 Failure Event:');
              console.log(JSON.stringify(event31341, null, 2));
            }

            resolve(status);
          }

          if (Date.now() - startTime > timeout) {
            clearInterval(interval);
            reject(new Error('Timeout waiting for session completion'));
          }

        } catch (error) {
          clearInterval(interval);
          reject(error);
        }
      }, 5000); // Check every 5 seconds
    });
  }
}

// Example usage
export async function runLightningExamples() {
  const lightning = new LightningOperations('http://localhost:8000');

  try {
    // Execute lift flow
    const liftResult = await lightning.executeLiftFlow(100000, 'gBTC');
    console.log(`\n🎯 Lift flow initiated. Session ID: ${liftResult.sessionId}`);

    // Monitor lift completion
    const liftStatus = await lightning.monitorSession(liftResult.sessionId);
    console.log(`\n📊 Lift final status: ${liftStatus.status}`);

    // Execute land flow
    const landResult = await lightning.executeLandFlow(
      50000,
      'lnbc500000n1p3k8...',
      'gBTC'
    );
    console.log(`\n🎯 Land flow initiated. Session ID: ${landResult.sessionId}`);

    // Monitor land completion
    const landStatus = await lightning.monitorSession(landResult.sessionId);
    console.log(`\n📊 Land final status: ${landStatus.status}`);

  } catch (error) {
    console.error('❌ Error:', error);
  }
}